cookbook/src/tx_segwit-v0.md

```diff
@@ -0,0 +1,342 @@
+   This will lock the output to the receiver's address.
+
+In `let change = TxOut {...}` we are instantiating the change output.
+It is very similar to the `spend` output, but we are now using the `const CHANGE_AMOUNT` that we defined earlier.
+And also we are locking the output to an address that we control:
+the `wpkh` public key hash that we generated earlier.
+This is done by setting the `script_pubkey` field to [`ScriptBuf::new_v0_p2wpkh(&wpkh)`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/script/struct.ScriptBuf.html#method.new_v0_p2wpkh),
+which generates P2WPKH-type of script pubkey.
+
+In `let unsigned_tx = Transaction {...}` we are instantiating the transaction we want to sign and broadcast using the [`Transaction`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Transaction.html) struct.
+We set the following fields:
+
+- `version` is the transaction version; it is a [`i32`](https://doc.rust-lang.org/std/primitive.u32.html) type.
+   We are using version `2` which means that [BIP68](https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki) applies.
+- `lock_time` is the transaction lock time;
+   it is a [`LockTime`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/locktime/absolute/enum.LockTime.html) enum.
+   We are using the [`from_height`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/locktime/absolute/enum.LockTime.html#method.from_height) method to set the lock time to the height `0`.
+   This will make the transaction valid immediately.
+- `input` is the input vector; it is a [`Vec<TxIn>`](https://doc.rust-lang.org/std/vec/struct.Vec.html) type.
+   We are using the `input` variable that we defined earlier wrapped in the [`vec!`](https://doc.rust-lang.org/std/macro.vec.html) macro for convenient initialization.
+- `output` is the output vector; it is a [`Vec<TxOut>`](https://doc.rust-lang.org/std/vec/struct.Vec.html) type.
+   We are using the `spend` and `change` variables that we defined earlier wrapped in the [`vec!`](https://doc.rust-lang.org/std/macro.vec.html) macro for convenient initialization.
+
+In `let mut sighash_cache = SighashCache::new(unsigned_tx);` we are instantiating a [`SighashCache`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html) struct.
+This is a type that ffficiently calculates [signature hash message](https://developer.bitcoin.org/devguide/transactions.html?highlight=sighash_all#signature-hash-types) for legacy, segwit and taproot inputs.
+We are using the `new` method to instantiate the struct with the `unsigned_tx` that we defined earlier.
+`new` takes any `Borrow<Transaction>` as an argument.
+[`Borrow<T>`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html) is a trait that allows us to pass either a reference to a `T` or a `T` itself.
+Hence, you can pass a `Transaction` or a `&Transaction` to `new`.
```

Comment:
Interesting, I would have expected this to mention `BorrowMut` but for some reason the code works without a mutable borrow - I don't actually understand how we are calling `tx.borrow_mut()` when the tx is only `Borrow` - bizzare. No action required, just mentioning.

Reply:
I think that
```rust
let mut sighash_cache = SighashCache::new(unsigned_tx);
```

actually takes ownership of `unsigned_tx`

We are not calling borrow under the hood,
see: https://docs.rs/bitcoin/latest/src/bitcoin/crypto/sighash.rs.html#557-559

Reply:
> actually takes ownership of `unsigned_tx`

Correct but the tx is of type `R: Borrow<Transaction>`

> We are not calling borrow under the hood,

On line 1080 of the same page you link to we call `self.tx.borrow_mut()`

https://docs.rs/bitcoin/latest/src/bitcoin/crypto/sighash.rs.html#1079